### ISSUE TITLE
No mechanism to notify callers that gazebo has disconnected

### ISSUE THREAD

**Issue by virtuald** [jpieper/pygazebo#7]
I'm having a bit of trouble following the logic, but basically in pygazebo.py at `_Connection::handle_read_raw_header()`, if the result is an empty string/byte array, then that means the connection has disconnected. At the moment, it just shows up as a ParseError, which is confusing to the user. 

At the moment I'm changing it to a DisconnectError, but I'm not convinced this is the right solution either -- but at least what happened was clearer.

**Comment by virtuald** [comment]
I also wonder why you didn't use asyncio/trollius protocol/transport stuff instead of using the low level socket transport.

**Comment by jpieper** [comment]
First, there were a couple of reasons for not using the protocol/transport stuff.  The primary one being that I didn't take the time to learn it!  Second, skimming it now, it looks like that without coroutines (and the Stream\* variants), it isn't clear that a protocol would necessarily be that much less verbose or clearer than the chained callback approach.

I explicitly didn't use coroutines anywhere in pygazebo, because I'm not aware of a syntax which is portable between python2 and python3.

If you think using the protocols transports would make things clearer and more robust, I'd be happy to review patches making it so.

**Comment by virtuald** [comment]
FYI

I misinterpreted the 'right' way to do I/O in asyncio.. apparently protocols are best left to asyncio, and instead there are higher level objects that one can use (namely StreamReader/StreamWriter).

I've refactored to use this stuff instead of using a raw socket, but unfortunately it does not perform as well as I would like it to. Profiling suggests that coroutines are the culprit -- since I have to convert them all to futures, there's quite a bit of overhead there that isn't present in the raw socket approach. 

One thing that profiling showed was that get_event_loop is called in a lot of places underneath the hood, so it's better to store the loop and pass it to the various asyncio functions that need it (namely Futures) instead of having things call get_event_loop() all over the place.
